### ISSUE TITLE
How to configure SQLitePCLRaw.bundle_sqlcipher as a plug in in a .Net Standard project?

### ISSUE THREAD

**Issue by BrianDT** [ericsink/SQLitePCL.raw#169]
Currently in the process of transitioning a number of PCL projects that use Microsoft’s SQLitePCL nuget package into .Net Standard projects that use sqlite-net-pcl. One reason sqlite-net-pcl was chosen was the ability to work with SQLitePCLRaw.bundle_sqlcipher as a move to data encryption was also a goal.
If both the SQLitePCLRaw.bundle_sqlcipher and sqlite-net-pcl nuget packages are added to a .Net Standard project this ultimately results in the following error when the UWP solution is built.

Error		Payload contains two or more files with the same destination path 'SQLitePCLRaw.batteries_v2.dll'. Source files: 
C:\Users\xxxx\.nuget\packages\SQLitePCLRaw.bundle_green\1.1.5\lib\uap10.0\SQLitePCLRaw.batteries_v2.dll
C:\Users\xxxx\.nuget\packages\SQLitePCLRaw.bundle_sqlcipher\1.1.7\lib\portable-net45+netcore45+wpa81+MonoAndroid10+MonoTouch10+Xamarin.iOS10\SQLitePCLRaw.batteries_v2.dll	SQLiteCryptTestApp			

Is there any way of controlling the dependencies in the package referencing in the .Net Standard project .csproj file
  <ItemGroup>
    <PackageReference Include="sqlite-net-pcl" Version="1.4.118" />
    <PackageReference Include="SQLitePCLRaw.bundle_sqlcipher" Version="1.1.7" />
  </ItemGroup>

That would manage the source of the SQLitePCLRaw.batteries_v2.dll?

**Comment by ericsink** [comment]
If you want to use bundle_sqlcipher with sqlite-net-pcl, the reference to bundle_sqlcipher must be placed in your app project, not in the same project as sqlite-net-pcl.

See this wiki page:

https://github.com/ericsink/SQLitePCL.raw/wiki/How-to-use-SQLCipher-with-SQLite-net

Unfortunately, you have another problem waiting for you after this one:  My bundle_sqlcipher package does not yet support UWP.  :-(

**Comment by BrianDT** [comment]
Hi Eric,
Thanks for the clarification.
I can see that this configuration builds correctly now that I have built the Android and iOS heads, shame about the UWP, any plans for supporting UWP in the future?
RegardsBrian on behalf of Visual Software Systems Ltd
Office 01276 63850
Mobile 07701087075
Fax 01276 679305
email brian@vsoftsys.co.uk This email is intended solely for the addressee(s) and the information it contains is confidential. If you are not the intended recipient;  (a) please delete this email and inform the sender as soon as possible, and  (b) any copying, distribution or other action taken or omitted to be taken in reliance upon it is prohibited and may be unlawful.   

    On Tuesday, 1 August 2017, 14:26, Eric Sink <notifications@github.com> wrote:
 

 If you want to use bundle_sqlcipher with sqlite-net-pcl, the reference to bundle_sqlcipher must be placed in your app project, not in the same project as sqlite-net-pcl.See this wiki page:https://github.com/ericsink/SQLitePCL.raw/wiki/How-to-use-SQLCipher-with-SQLite-netUnfortunately, you have another problem waiting for you after this one: My bundle_sqlcipher package does not yet support UWP. :-(—

**Comment by ericsink** [comment]
Yes, I do intend to include a UWP sqlcipher build. Requests for it are getting more frequent.

**Comment by ericsink** [comment]
Release 1.1.9 included support for sqlcipher with UWP.

**Comment by 1iveowl** [comment]
I'm having issues with 1.1.9 and UWP.

I'm using Microsoft.NETCore.UniversalWindowsPlatform 6.0.5.

And I get this error:

> Severity	Code	Description	Project	File	Line	Suppression State
Error		Payload contains two or more files with the same destination path 'SQLitePCLRaw.batteries_v2.dll'. Source files: 
C:\Users\jaspe\.nuget\packages\sqlitepclraw.bundle_green\1.1.5\lib\uap10.0\SQLitePCLRaw.batteries_v2.dll
C:\Users\jaspe\.nuget\packages\sqlitepclraw.bundle_sqlcipher\1.1.9\lib\uap10.0\SQLitePCLRaw.batteries_v2.dll

Any help on what is going on here?

This is a Xamarin Project running .NET Standard 2.0. I'm using `sqlite-net-pcl`, which I've installed in the main .NET Standard library. the UWP, iOS and Android libraries I've installed `sqlitepclraw.bundle_sqlcipher`. 

iOS and Android compiles juet fine. I'm only having issues with UWP.

**Comment by ericsink** [comment]
@liveowl Hmmm.

Here is something to try, just to see if it eliminates the problem:

Instead of sqlite-net-pcl, use sqlite-net-base, which is the same package, except without a bundle reference.

This will require you to add a reference to bundle_sqlcipher in each of your app projects, iOS, Android, and UWP.

**Comment by 1iveowl** [comment]
I tried this. And it worked.

Also, I like the simplicity of the approach. 

sqlite-net-base only seen to be available in beta. Not a great issue, just an observation.

Thank you!

**Comment by 1iveowl** [comment]
Sorry. I was too fast. 

Yes, this now works in UWP, but I get this error when trying to use Andriod:

> Exception: Exception has been thrown by the target of an invocation., Inner: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = SettingsViewModel (ReflectionActivator), Services = [AXExpense2018.ViewModel.SettingsViewModel], Lifetime = Autofac.Core.Lifetime.CurrentScopeLifetime, Sharing = None, Ownership = OwnedByLifetimeScope ---> An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = EntityService (DelegateActivator), Services = [IEcryptedDB.IEntityService], Lifetime = Autofac.Core.Lifetime.RootScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope ---> You need to call SQLitePCL.raw.SetProvider().  If you are using a bundle package, this is done by calling SQLitePCL.Batteries.Init(). (See inner exception for details.) (See inner exception for details.)

This is a Xamarin project running .NET Standard 2.0

**Comment by ericsink** [comment]
Indeed, I forgot to mention the need to call `SQLitePCL.Batteries_V2.Init()` as well.

**Comment by 1iveowl** [comment]
FYI. 

Ran into a related issue when using [Microsoft AppCenter](http://appcenter.ms) in a UWP project. See details [here](https://github.com/Microsoft/AppCenter-SDK-DotNet/issues/569#issuecomment-356138645)

**Comment by ChaseCarlile** [comment]
Thanks, @1iveowl .  That one had me blocked for a while.  I set up a repo here to demonstrate the problem and fix:  https://github.com/ChaseCarlile/sqlcipher_AppCenter_Issue .  

TLDR - Add this to your UWP csproj:

    <PackageReference Include="sqlite-net-pcl">
      <Version>1.*</Version>
      <ExcludeAssets>All</ExcludeAssets>
    </PackageReference>
    <PackageReference Include="SQLitePCLRaw.bundle_green">
      <Version>1.*</Version>
      <ExcludeAssets>All</ExcludeAssets>
    </PackageReference>

**Comment by Samrak** [comment]
hello, what kind of secure algorithm is used? aes-128 or aes-256 ?
Could I take enough information about its security background?

**Comment by ericsink** [comment]
@Samrak SQLCipher uses aes 256:

https://www.zetetic.net/sqlcipher/
